File: utils/matlab_utils/process_database_liver.m
addpath('../../tools/nifti_library/');

%folder with .nii files 
niftis_path = '../../Database/media/nas/01_Datasets/CT/LITS/Training Batch 1/';

root_process_database = '../../LiTS_database/';
folder_volumes = strcat(root_process_database, 'images_volumes/');
folder_seg_liver = strcat(root_process_database, 'liver_seg/');
folder_seg_item = strcat(root_process_database, 'item_seg/');


if ~(exist(root_process_database, 'dir'))
    mkdir(root_process_database);
end
if ~(exist(folder_volumes, 'dir'))
    mkdir(folder_volumes);
end
if ~(exist(folder_seg_liver, 'dir'))
    mkdir(folder_seg_liver);
end
if ~(exist(folder_seg_item, 'dir'))
    mkdir(folder_seg_item);
end

files_dir = dir(niftis_path);
copy_files_dir = files_dir;
filenames = [];

list_file_names = [];

% just keep those files that belong the volumes or segmentations
for k = 1:(length(files_dir))
	if (files_dir(k).name(1) == 'v' || files_dir(k).name(1) == 's' )
        filenames = [filenames files_dir(k)];
	end
end


for l = 1:(length(filenames))
	name = filenames(l).name;
    if (name(1) == 'v')
        display(['Processing Volume'])
        path_file = strcat(niftis_path, name);
        index = strfind(name,'.nii');
        folder_volume = strcat(folder_volumes, name((8):index-1));
        volume = load_nii(path_file);
        imgs = volume.img;
        imgs(imgs<-150) = -150;
        imgs(imgs>250) = 250;
        imgs = single(imgs);
        img_volume = 255*(imgs - min(imgs(:)))/(max(imgs(:)) - min(imgs(:)));
        if ~(exist(folder_volume, 'dir'))
            mkdir(folder_volume);
        end
		for j=1:(size(img_volume,3))
			section = img_volume(:,:,j);
            filename_for_section = strcat(folder_volume, '/', num2str(j), '.mat');
            save(filename_for_section, 'section', '-v6');
        end
    elseif (name(1) == 's')
        display(['Processing Segmentation'])
        path_file = strcat(niftis_path, name);
        index = strfind(name,'.nii');
        folder_seg_item_num = strcat(folder_seg_item, name(14:index-1));
        folder_seg_liver_num = strcat(folder_seg_liver, name(14:index-1));
        segmentation = load_nii(path_file);
        img_seg = uint8(segmentation.img);
        img_seg_item = img_seg;
        img_seg_liver = img_seg;
        img_seg_item(img_seg_item == 1)=0;
        img_seg_item(img_seg_item == 2)=1;
        img_seg_liver(img_seg_liver == 2)=1;
        if ~(exist(folder_seg_item_num, 'dir'))
            mkdir(folder_seg_item_num);
        end
        if ~(exist(folder_seg_liver_num, 'dir'))
            mkdir(folder_seg_liver_num);
        end
        for j=1:(size(img_seg_item,3))
            item_seg_section = img_seg_item(:,:,j)*255;
            liver_seg_section = img_seg_liver(:,:,j)*255;
            filename_for_seg_item_section = strcat(folder_seg_item_num, '/', num2str(j), '.png');
            filename_for_seg_liver_section = strcat(folder_seg_liver_num, '/', num2str(j),  '.png');
            imwrite(item_seg_section, filename_for_seg_item_section);
            imwrite(liver_seg_section, filename_for_seg_liver_section);
        end
    end
end

File: utils/matlab_utils/process_database_liver.py
import os
import nibabel as nib # pip install nibabel 
# https://nipy.org/nibabel/nifti_images.html

import numpy as np
import matplotlib.pyplot as plt
import scipy.io
from PIL import Image


# path constants
niftis_path = '../../Database/media/nas/01_Datasets/CT/LITS/Training Batch 1/'
root_process_database = '../../LiTS_database/'

niftis_path = r'C:\Users\grego\OneDrive\Projects\Python Projects\fake_niftis'
root_process_database = r'C:\Users\grego\OneDrive\Projects\Python Projects\fake_root'

folder_volumes = os.path.join(root_process_database, 'images_volumes/')
folder_seg_liver = os.path.join(root_process_database, 'liver_seg/')
folder_seg_item = os.path.join(root_process_database, 'item_seg/')


# create non-existent paths
folder_paths = [root_process_database, folder_volumes, folder_seg_liver, folder_seg_item]
    
for p in folder_paths:
    if not os.path.exists(p):
        os.mkdir(p)


# filter to only files starting with v (volume) or s (segmentation)
filenames = [filename for filename in os.listdir(niftis_path) if filename[0] in ('v', 's')]


for filename in filenames:
    path_file = os.path.join(niftis_path, filename)
    index = filename.find('.nii') + 1 # +1 to account for matlab --> python
    if filename[0] == 'v':
        print(f'Processing Volume {filename}')
        folder_volume = os.path.join(folder_volumes, filename[7:index-1])
        volume = nib.load(path_file) # load nifti
        imgs = volume.get_fdata() # get 3d NumPy array

        # clipping
        imgs[imgs<-150] = -150
        imgs[imgs>250] = 250

        imgs = imgs.astype(np.float32) # equivalent to matlab single()
        img_max, img_min = (np.max(imgs), np.min(imgs))

        # create folder_volume folder
        img_volume = 255*(imgs - img_min)/(img_max-img_min)
        if not os.path.exists(folder_volume):
            os.mkdir(folder_volume)
        
        for k in range(img_volume.shape[2]):
            section = img_volume[:,:,k]
            filename_for_section = os.path.join(folder_volume, str(k+1) + '.mat')
            scipy.io.savemat(filename_for_section, {'section': section})
    else:
        print(f'Processing Segmentation {filename}')
        folder_seg_item_num = os.path.join(folder_seg_item, filename[13:index-1])
        folder_seg_liver_num = os.path.join(folder_seg_liver, filename[13:index-1])
        segmentation = nib.load(path_file)
        img_seg = segmentation.get_fdata().astype(np.uint8)
        print(img_seg.shape)
        # binarize and normalize data
        img_seg_item = img_seg.copy()
        img_seg_liver = img_seg.copy()

        # create masks
        img_seg_item[img_seg_item == 1] = 0
        img_seg_item[img_seg_item == 2] = 1
        img_seg_liver[img_seg_liver == 2] = 1

        # create dirs
        if not os.path.exists(folder_seg_item_num):
            os.mkdir(folder_seg_item_num)
        if not os.path.exists(folder_seg_liver_num):
            os.mkdir(folder_seg_liver_num)
        
        # save images
        for k in range(1, img_seg_item.shape[2]):

            # item
            item_seg_section = np.fliplr(np.flipud(img_seg_item[:,:,k]*255)) # flip on both axes
            item_seg_filename = os.path.join(folder_seg_item_num, str(k+1) + '.png')
            im_item = Image.fromarray(item_seg_section)
            im_item.save(item_seg_filename)
            
            # liver
            liver_seg_section = np.fliplr(np.flipud(img_seg_liver[:,:,k]*255))
            liver_seg_filename = os.path.join(folder_seg_liver_num, str(k+1) + '.png')
            im_liver = Image.fromarray(liver_seg_section)
            im_liver.save(liver_seg_filename)
